### PR TITLE
Experimental: Add `BasisLieHighestWeight`

### PR DESCRIPTION
**Pull request by lgoettgens**
This is another try to https://github.com/oscar-system/Oscar.jl/pull/2115 and https://github.com/oscar-system/Oscar.jl/pull/2402 (now on my fork so I can manage access rights myself).

Background: This is code written mainly by @BenWilop (student assistant in Aachen) that is needed for and referenced in the book chapter of @gfourier and Xin Fang. I will be helping them a bit to "oscarify" everything.

@BenWilop please continue working on this branch. If you encounter any access rights issues, please write me an email.

~@ everyone please wait with reviews and comments for a bit, this PR is currently just to enable collaboration.~

### CODE REVIEW

**Comment by lgoettgens** [comment]
~This is not done yet. There are some printing issues still to be addressed, and the docstrings have to be made consistent with the interface below.~
But since this code is used for the OSCAR book, I would like to give some time for comments.

~Please do not merge yet.~

**Comment by lgoettgens** [comment]
Some ToDos for the future (not in this PR):

- [ ] Merge the `LieAlgebraStructure` here with Lie algebras from root systems, once they have all the needed functionality
- [ ] Add docs
- [ ] Excise MBOld.jl and replace with better property-based testing, or serialize some test cases to test against

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 1]
```diff
@@ -0,0 +1,408 @@
+@doc raw"""
+    basis_lie_highest_weight_operators(type::Symbol, rank::Int)
+
+Lists the operators available for a given simple Lie algebra of type `type_rank`,
+together with their index.
+Operators of the form $f_\alpha$ are shown as the coefficients of $\alpha$
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
Operators $f_\alpha$ of negative roots are shown as the coefficients of the corresponding positive root.
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+@doc raw"""
+    basis_lie_highest_weight(type::Symbol, rank::Int, highest_weight::Vector{Int}; monomial_ordering::Symbol=:degrevlex)
+    basis_lie_highest_weight(type::Symbol, rank::Int, highest_weight::Vector{Int}, birational_sequence::Vector{Int}; monomial_ordering::Symbol=:degrevlex)
+    basis_lie_highest_weight(type::Symbol, rank::Int, highest_weight::Vector{Int}, birational_sequence::Vector{Vector{Int}}; monomial_ordering::Symbol=:degrevlex)
+
+Computes a monomial basis for the highest weight module with highest weight
+`highest_weight` (in terms of the fundamental weights $\omega_i$),
+for a simple Lie algebra of type `type_rank`.
+
+If no birational sequence is given, all operators in the order of `basis_lie_highest_weight_operators` are used.
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
If no birational sequence is specified, all operators in the order of `basis_lie_highest_weight_operators` are used.
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+
+# Examples
+```jldoctest
+julia> base = BasisLieHighestWeight.basis_lie_highest_weight(:A, 2, [1, 1])
+Monomial basis of a highest weight module
+  of highest weight [1, 1]
+  of dimension 8
+  with monomial ordering degrevlex([x1, x2, x3])
+over Lie Algebra of type A2
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+    [1, 0]
+    [0, 1]
+
+julia> base = BasisLieHighestWeight.basis_lie_highest_weight(:A, 3, [2, 2, 3]; monomial_ordering = :lex)
+Monomial basis of a highest weight module
+  of highest weight [2, 2, 3]
+  of dimension 1260
+  with monomial ordering lex([x1, x2, x3, x4, x5, x6])
+over Lie Algebra of type A3
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+    [0, 1, 0]
+    [0, 0, 1]
+
+julia> base = BasisLieHighestWeight.basis_lie_highest_weight(:A, 2, [1, 0], [1,2,1])
+Monomial basis of a highest weight module
+  of highest weight [1, 0]
+  of dimension 3
+  with monomial ordering degrevlex([x1, x2, x3])
+over Lie Algebra of type A2
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where  the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+  and the basis was generated by Minkowski sums of the bases of the following highest weight modules:
+    [1, 0]
+
+julia> base = BasisLieHighestWeight.basis_lie_highest_weight(:A, 2, [1, 0], [[1,0], [0,1], [1,0]])
+Monomial basis of a highest weight module
+  of highest weight [1, 0]
+  of dimension 3
+  with monomial ordering degrevlex([x1, x2, x3])
+over Lie Algebra of type A2
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where  the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+  and the basis was generated by Minkowski sums of the bases of the following highest weight modules:
+    [1, 0]
+
+julia> base = BasisLieHighestWeight.basis_lie_highest_weight(:C, 3, [1, 1, 1]; monomial_ordering = :lex)
+Monomial basis of a highest weight module
+  of highest weight [1, 1, 1]
+  of dimension 512
+  with monomial ordering lex([x1, x2, x3, x4, x5, x6, x7, x8, x9])
+over Lie Algebra of type C3
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where  the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+
+# Examples
+```jldoctest
+julia> base = BasisLieHighestWeight.basis_lie_highest_weight_lusztig(:D, 4, [1,1,1,1], [4,3,2,4,3,2,1,2,4,3,2,1])
+Monomial basis of a highest weight module
+  of highest weight [1, 1, 1, 1]
+  of dimension 4096
+  with monomial ordering wdegrevlex([x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12], [1, 1, 3, 2, 2, 1, 5, 4, 3, 3, 2, 1])
+over Lie Algebra of type D4
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where  the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+
+# Examples
+```jldoctest
+julia> BasisLieHighestWeight.basis_lie_highest_weight_string(:B, 3, [1,1,1], [3,2,3,2,1,2,3,2,1])
+Monomial basis of a highest weight module
+  of highest weight [1, 1, 1]
+  of dimension 512
+  with monomial ordering neglex([x1, x2, x3, x4, x5, x6, x7, x8, x9])
+over Lie Algebra of type B3
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where  the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+    [0, 1, 0]
+    [0, 0, 1]
+
+julia> BasisLieHighestWeight.basis_lie_highest_weight_string(:A, 4, [1,1,1,1], [4,3,2,1,2,3,4,3,2,3])
+Monomial basis of a highest weight module
+  of highest weight [1, 1, 1, 1]
+  of dimension 1024
+  with monomial ordering neglex([x1, x2, x3, x4, x5, x6, x7, x8, x9, x10])
+over Lie Algebra of type A4
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where  the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+      
+# Examples
+```jldoctest
+julia> BasisLieHighestWeight.basis_lie_highest_weight_pbw(:A, 3, [1,1,1])
+Monomial basis of a highest weight module
+  of highest weight [1, 1, 1]
+  of dimension 64
+  with monomial ordering neglex([x1, x2, x3, x4, x5, x6])
+over Lie Algebra of type A3
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+
+# Examples
+```jldoctest
+julia> BasisLieHighestWeight.basis_lie_highest_weight_nz(:C, 3, [1,1,1], [3,2,3,2,1,2,3,2,1])
+Monomial basis of a highest weight module
+  of highest weight [1, 1, 1]
+  of dimension 512
+  with monomial ordering degrevlex([x1, x2, x3, x4, x5, x6, x7, x8, x9])
+over Lie Algebra of type C3
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where  the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

File: experimental/BasisLieHighestWeight/src/UserFunctions.jl [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+    [0, 1, 0]
+    [0, 0, 1]
+
+julia> BasisLieHighestWeight.basis_lie_highest_weight_nz(:A, 4, [1,1,1,1], [4,3,2,1,2,3,4,3,2,3])
+Monomial basis of a highest weight module
+  of highest weight [1, 1, 1, 1]
+  of dimension 1024
+  with monomial ordering degrevlex([x1, x2, x3, x4, x5, x6, x7, x8, x9, x10])
+over Lie Algebra of type A4
+  where the birational sequence used consists of operators to the following weights (given as coefficients w.r.t. alpha_i):
```
Comment (gfourier) [inline experimental/BasisLieHighestWeight/src/UserFunctions.jl]:
```suggestion
  where the used birational sequence consists of the following roots (given as coefficients w.r.t. alpha_i):
```

**Review by gfourier — APPROVED** [review]

File: experimental/BasisLieHighestWeight/src/LieAlgebras.jl
Comment (micjoswig) [inline experimental/BasisLieHighestWeight/src/LieAlgebras.jl:1]: OK, there is an experimental `LieAlgebra` type, and now there is also `LieAlgebraStructure`.  This looks a bit weird to me.  Please enlighten me about your goals.

**Review by micjoswig — COMMENTED** [review]
The Demazure stuff needs to be compared and probably unified with the existing function demazure_character, which already exists in the polyhedral section of the source code.

I had just discussed with @tbrysiewicz that it would make sense to move that function near the algebra or group theory sections.

Calling polymake through the `Polymake.jl` interface is undesirable, unless there is a specific reason.  Instead you should use the existing interface which is explained, e.g.,  [in this OSCAR book chapter](https://arxiv.org/abs/2308.07459).  This should also simplify your code, e.g., because there is no need for you to explicitly homogenize.  You might want to talk to @benlorenz .

**Comment by gfourier** [comment]
> The Demazure stuff needs to be compared and probably unified with the existing function demazure_character, which already exists in the polyhedral section of the source code.

Yes, this Demazure stuff has to be generalized (beyond type A as in the polyhedral section) and then maybe moved to a Lie algebra section.

File: experimental/BasisLieHighestWeight/src/LieAlgebras.jl
Comment (lgoettgens) [reply experimental/BasisLieHighestWeight/src/LieAlgebras.jl:1]: `LieAlgebraStructure` is only there to exist temporarily (as a thin wrapper around), as `LieAlgebra` currently does not support everything that is needed here. As already noted in https://github.com/oscar-system/Oscar.jl/pull/2936#issuecomment-1794576507, this is on the todo-list to merge. But as that involves more work of different people (e.g. @felix-roehrich), and is mostly unrelated to the stuff here, I would like to move that piece of work to a future PR.

**Comment by lgoettgens** [comment]
> The Demazure stuff needs to be compared and probably unified with the existing function demazure_character, which already exists in the polyhedral section of the source code.
> 
> I had just discussed with @tbrysiewicz that it would make sense to move that function near the algebra or group theory sections.
> 
> Calling polymake through the `Polymake.jl` interface is undesirable, unless there is a specific reason. Instead you should use the existing interface which is explained, e.g., [in this OSCAR book chapter](https://arxiv.org/abs/2308.07459). This should also simplify your code, e.g., because there is no need for you to explicitly homogenize. You might want to talk to @benlorenz .

I am a bit confused about what you are talking about here @micjoswig. Could it be that you are looking at an old revision of this? In the current form of this PR, all Demazure functionality has been purged. Furthermore, the code calling polymake has already been changed more than a week ago (to this: https://github.com/oscar-system/Oscar.jl/pull/2936/files#diff-5cb6cccdda2ba91185fae3c9bd3d4dc0e0ddd32b4e8f3044dc3c622ce8f60f2fR89).

**Comment by micjoswig** [comment]
OK, probably I looked at the wrong thing.  Pull requests from forks continue to confuse me.

**Comment by micjoswig** [comment]
OK, I don't understand all the details, e.g., for lack of GAP background.  But essentially this looks alright for now.

The best way of dealing with Dynkin diagrams (e.g., in terms of interfaces) should probably be discussed with @ulthiel .

One more thing: please employ `@docs` to make the existing documentation also available in the generated HTML docs.  Can't hurt to throw in a few more words of explanations.

**Comment by lgoettgens** [comment]
I just added all docstrings to a documentation page. Could we get this merged once CI succeeds? Or are there any other objections?

**Review by micjoswig — APPROVED** [review]

**Pull request merged**